### PR TITLE
POC for updates to x_ref preprocessing args

### PR DESCRIPTION
**Pull request by ascillitoe**
This PR is a proof of concept for possible changes to the drift detector kwargs for reference data preprocessing. Changes are only made to the `KSDrift` detector. 

### Key changes

The `preprocess_x_ref` kwarg is replaced by two kwargs:
1. `x_ref_preprocess`: Signals whether the provided `x_ref` has already been preprocessed or not. If `True`, then `preprocess_fn` is only applied to the test data.
2. `preprocess_at_init`: This has the same meaning as the old `preprocess_x_ref` i.e. if `True`, `x_ref` is preprocessed at detector instantiation, otherwise preprocessing is performed when `predict` is called. 

Internally the two kwargs (and `preprocess_fn`) are used to set `self.preprocess_at_init` and `self.preprocess_at_pred`, which independently flag whether to perform preprocessing of `x_ref` at init and predict. 

The above kwargs are hopefully more clear to the user, as they draw a clearer distinction between whether or not to preprocess `x_ref`, and when to preprocess `x_ref`. `x_ref_preprocess` also simplifies ongoing work on config based saving/loading.

### Tests
This has only been tested on the `cd_ks_cifar10` notebook.

### CODE REVIEW

**Comment by jklaise** [comment]
@ascillitoe after reading https://stackoverflow.com/questions/55377831/difference-between-deprecationwarning-pendingdeprecationwarning-and-futurewarni I think I was mistaken (or at least misguided). It appears that as library developers our "users" would be primarily application/library developers, so it appears that `DeprecationWarning` would be appropriate. That being said I don't think it's a huge deal.

**Comment by ascillitoe** [comment]
@jklaise, ah interesting, thanks for that. I am getting back to this today so will add a `DeprecationWarning`.

File: alibi_detect/cd/ks.py [hunk 1]
```diff
@@ -51,10 +59,16 @@ def __init__(
         data_type
             Optionally specify the data type (tabular, image or time-series). Added to metadata.
         """
+        if 'preprocess_x_ref' in kwargs:  # Note - to be removed in the future
+            preprocess_at_init = kwargs['preprocess_x_ref']
+            warning_msg = "The `preprocess_x_ref` kwarg has ben replaced by `preprocess_at_init`, and " \
+                          "will be removed in a future version."
+            warnings.warn(warning_msg, DeprecationWarning)
```
Comment (ascillitoe) [inline alibi_detect/cd/ks.py:66]: @jklaise any thoughts on this? Seems a bit ugly to add this to every drift detector, but I'm a bit wary about adding `**kwargs` to the base classes instead...

Comment (jklaise) [reply alibi_detect/cd/ks.py:66]: could we implement this with a decorator? e.g. something like

Comment (ascillitoe) [reply alibi_detect/cd/ks.py]: Thanks @jklaise, this looks like a nice solution.

**Comment by ascillitoe** [comment]
Closing this in order to open an upstream PR.

**Pull request closed without merging**